DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/merge_sort.cpp

```diff
@@ -0,0 +1,334 @@
+
+// determines whether we will use USM host or device allocations to move data
+// between host and the device.
+#if defined(USM_HOST_ALLOCATIONS)
+constexpr bool kUseUSMHostAllocation = true;
+#else
+constexpr bool kUseUSMHostAllocation = false;
+#endif
+
+// the number of merge units, must be a power of 2
+#ifndef MERGE_UNITS
+#ifdef FPGA_EMULATOR
+#define MERGE_UNITS 4
+#else
+#define MERGE_UNITS 16
+#endif
+#endif
+constexpr size_t kMergeUnits = MERGE_UNITS;
+
+// sanity check on number of merge units
+static_assert(kMergeUnits > 0);
+static_assert(IsPow2(kMergeUnits));
+
+////////////////////////////////////////////////////////////////////////////////
+// Forward declare functions used in main
+template<typename ValueT, typename IndexT, typename KernelPtrType>
+double fpga_sort(queue& q, ValueT *in_vec, ValueT *out_vec, IndexT count);
+
+template<typename T>
+bool validate(T *val, T *ref, unsigned int count);
+////////////////////////////////////////////////////////////////////////////////
+
+// main
+int main(int argc, char* argv[]) {
+  // type to sort, needs a compare function!
+  using ValueT = int;
+
+  // the type used to index in the sorter
+  using IndexT = unsigned int;
+
+  bool passed = true;
+
+#ifdef FPGA_EMULATOR
+  IndexT count = 32;
+  int runs = 1;
+#else
+  IndexT count = 1 << 24;
+  int runs = 17;
+#endif
+
+  // get the size of the input as the first command line argument
+  if(argc > 1) {
+    count = atoi(argv[1]);
+  }
+
+  // get the number of runs as the second command line argument
+  if (argc > 2) {
+    runs = atoi(argv[2]);
+  }
+
+  // enforce at least two runs
+  runs = std::max((int)2, runs);
+  
+  // check args
+  if (count <= kMergeUnits) {
+    std::cerr << "ERROR: count must be greater than number of merge units\n";
+    std::terminate();
+  } else if (count > std::numeric_limits<IndexT>::max()) {
+    std::cerr << "ERROR: the index type does not have bits to count to "
+              << "'count'\n";
+    std::terminate();
+  }
+
+  // the device selector
+#ifdef FPGA_EMULATOR
+  INTEL::fpga_emulator_selector selector;
+#else
+  INTEL::fpga_selector selector;
+#endif
+
+  // create the device queue
+  queue q(selector, dpc_common::exception_handler);
+
+  // make sure the device supports USM device allocations
+  device d = q.get_device();
+  if (!d.get_info<info::device::usm_device_allocations>()) {
+    std::cerr << "ERROR: The selected device does not support USM device"
+              << " allocations\n";
+    std::terminate();
+  }
+
+  // make sure the device support USM host allocations if we chose to use them
+  if (!d.get_info<info::device::usm_host_allocations>() &&
+      kUseUSMHostAllocation) {
+    std::cerr << "ERROR: The selected device does not support USM host"
+              << " allocations\n";
+    std::terminate();
+  }
+
+  // the input and output data
+  std::vector<ValueT> in_vec(count), out_vec(count), ref(count);
+
+  // generate some random input data
+  srand(time(NULL));
```

Comment:
I suggest you make the seed a command line argument, so results are reproducible.  If you want to keep randomization based on time an option, you could do that if no command line seed is provided.

Reply:
Done.